File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {

public int lengthOfLongestSubstring(String s) {
    
    if(s.length() == 0)             //if string is empty then return 0
        return 0;
    
    HashMap<Character,Integer> map = new HashMap<>();       //help to get the previous index of repeating character
    
    int maxCount = Integer.MIN_VALUE;                       //Store minimum value in it, can take zero also
    int count=0;                                            //a counter to count the length 
    
    for(int i=0;i<s.length();i++)
    {
        if(map.containsKey(s.charAt(i))){                   //Map contains that char then store previous index of that repeating char in i by +1
            count=1;                                        //count reinitialise to 1 and delete all the elements from map 
            i=map.get(s.charAt(i))+1;                       //to find another longest substring
            map.clear();               
        }
        else
            count++;
        
        map.put(s.charAt(i),i);                             // add char in the map with its index value
        maxCount=Math.max(count,maxCount);                  //store the longest substring length
    }
    
    return maxCount;
    
}
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if(len(s) == 1):return 1
        
        i = 0
        j = 1
        siz = 0
        maxVal = 0
        while i < len(s) and j <= len(s):
            sub = s[i:j]
            siz = len(sub)
            if(j < len(s) and  not (s[j] in sub)):
                j+=1
            else:
                maxVal = max(siz, maxVal)
                i+=1
                j=i+1
                
        return max(maxVal, siz)